proposals/0448-type-variable-scoping.rst

```diff
@@ -198,42 +233,74 @@ All other variables are explicitly bound, and the inconsistency means more to le
-However, there is one exception to the unification of namespaces: lower case variables in type signatures bound "like terms" still are treated as free and implicitly bound with a ``forall`` instead::
+However, there is one exception to the unification of namespaces: lower case variables in type signatures bound "like terms" still are treated as free and implicitly bound instead::
 
   t = Int
-  x :: t -- sugar for 'forall t. t', not 't ~ Int'
+  x :: t -- sugar for 'forall t. t', not a use of 't' resolving to 'Int'
   x = 0
 
-Should the ``t`` in ``x :: t`` cause an implicit ``forall t.`` to be synthesized or not? With ``-XNoImplicitForAll``, we know
-it will not, and thus can refer to the ``t`` defined above, once such a reference is possible (left to another proposal).
+  t = Int
+  foo (x :: t) = 0 -- sugar for 'foo = let t = _ in \(x :: t) -> 0'
+
+Should the ``t`` in each ``x :: t`` cause implicit ``forall t.`` and ``let t = _ in`` to be synthesized or not?
 
-Motivation for ``-Wpattern-signature-binds``
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+With ``-XNoImplicitBinds``, we know it will not, and thus can refer to the ``t`` defined above, once such a reference is possible (left to another proposal).
 
-This is necessary in order to uphold the `Lexical Scoping Principle`_, part (a).
+Without ``-XImplicitBinds`` we have no choice but do the implicit desugaring that violates the unified namespace abstraction
+--- otherwise we would be *forklike* in the sense that rather than giving new meaning to previously invalid programs, we would be *changing* the meaning of previously *valid* programs.
+Put another way, ``-XImplicitBinds`` and any ``-XUnifiedNamespace`` must be mutually exclusive, so we need a way for Haskell programmers to opt out of ``-XImplicitBinds`` first.
 
 Proposed Change Specification
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 Points below up to and including the new (backward-compatible) definition of
-``-XScopedTypeVariables`` come from not-yet-accepted proposal `#238`_. The point
-about ``-XImplicitForAll`` is a restatement of (part of) accepted proposal `#285`_.
-The other part of `#285`_ has been modified to use ``-Wpattern-signature-binds``.
+``-XScopedTypeVariables`` come from not-yet-accepted proposal `#238`_.
+``-XImplicitBinds`` is a fixed and simplified (via combining extensions) version of accepted proposal `#285`_.
 
 1. Re-purpose deprecated extension ``-XPatternSignatures``. With ``-XPatternSignatures``, we
    allow type signatures in patterns. These signatures can mention in-scope
-   type variables as variable occurrences. A mention of an out-of-scope variable
-   binds the type variable as an alias of the type it is unified with (as today).
+   type variables as variable occurrences, but can not bind type variables without the
+   separate ``-XImplicitBinds`` extension. Do note that extension is on by default, however.
```

Comment:
.

Reply:
Isn't this more than a rewording? Before, implicit pattern signature binds were gated by `-XPatternSignatures` alone, but warned. Now, they are gated by the combination of `-XPatternSignatures` and `-XImplicitBinds`.

Yes, the difference in practice is subtle because the latter is on by default, but generally when I write these things I try to specify the effects of extensions and the defaulting of extensions separately. (Dealing with both at once I find confusing.)